cmd/mt-replicator-via-tsdb/metrics.go

```diff
@@ -0,0 +1,250 @@
+	config.Config.Version = sarama.V0_10_0_0
+
+	err := config.Validate()
+	if err != nil {
+		return nil, err
+	}
+	consumer, err := cluster.NewConsumer(srcBrokers, *group, []string{*srcTopic}, config)
+	if err != nil {
+		log.Error(3, "failed to connect to source brokers %v.", srcBrokers)
+		return nil, err
+	}
+	tsdbClient := &http.Client{
+		Timeout: time.Duration(10) * time.Second,
+	}
+
+	return &MetricsReplicator{
+		consumer:   consumer,
+		tsdbClient: tsdbClient,
+		tsdbKey:    *destinationKey,
+		tsdbUrl:    *destinationUrl,
+
+		shutdown:   make(chan struct{}),
+		writeQueue: make(chan *writeRequest, 10),
+	}, nil
+}
+
+func (r *MetricsReplicator) Consume() {
+	buf := make([]*schema.MetricData, 0)
+	accountingTicker := time.NewTicker(time.Second * 10)
+	flushTicker := time.NewTicker(time.Second)
+	counter := 0
+	counterTs := time.Now()
+	msgChan := r.consumer.Messages()
+
+	flush := func(topic string, partition int32, offset int64) {
+		mda := schema.MetricDataArray(buf)
+		data, err := msg.CreateMsg(mda, 0, msg.FormatMetricDataArrayMsgp)
+		if err != nil {
+			panic(err)
+		}
+		// this will block when the writeQueue fills up. This will happen if
+		// we are consuming at a faster rate then we can publish, or if publishing
+		// is failing for some reason.
+		r.writeQueue <- &writeRequest{
+			data:      data,
+			topic:     topic,
+			partition: partition,
+			offset:    offset,
+			count:     len(buf),
+		}
+		counter += len(buf)
+		buf = buf[:0]
+	}
+
+	var m *sarama.ConsumerMessage
+	var ok bool
+	defer func() {
+		close(r.writeQueue)
+		r.wg.Done()
+	}()
+	for {
+		select {
+		case m, ok = <-msgChan:
+			if !ok {
+				if len(buf) != 0 {
+					flush(m.Topic, m.Partition, m.Offset)
+				}
+				return
+			}
+			md := &schema.MetricData{}
+			_, err := md.UnmarshalMsg(m.Value)
+			if err != nil {
+				log.Error(3, "kafka-mdm decode error, skipping message. %s", err)
+				continue
+			}
+
+			buf = append(buf, md)
+			if len(buf) > *producerBatchSize {
+				flush(m.Topic, m.Partition, m.Offset)
+				// reset our ticker
+				flushTicker.Stop()
+				flushTicker = time.NewTicker(time.Second)
+			}
+		case <-flushTicker.C:
+			if len(buf) == 0 {
+				continue
+			}
+			flush(m.Topic, m.Partition, m.Offset)
+		case t := <-accountingTicker.C:
+			log.Info("%d metrics processed in last %.1fseconds.", counter, t.Sub(counterTs).Seconds())
+			counter = 0
+			counterTs = t
+		case <-r.shutdown:
+			if len(buf) > 0 {
+				flush(m.Topic, m.Partition, m.Offset)
+			}
+			return
+		}
+	}
+}
+
+func (r *MetricsReplicator) Stop() {
+	r.consumer.Close()
+	close(r.shutdown)
+	log.Info("Consumer closed.")
+	done := make(chan struct{})
+	go func() {
+		r.wg.Wait()
+		close(done)
+	}()
+
+	select {
+	case <-time.After(time.Minute):
+		log.Info("shutdown not complete after 1 minute. Abandoning inflight data.")
+	case <-done:
+		log.Info("shutdown complete.")
+	}
+	return
+}
+
+func (r *MetricsReplicator) Start() {
```

Comment:
i would reorder the code. after the constructor put Start, then Stop, then consume and flush (make the latter two non-exported), the order seems pretty arbitrary now, I think it's generally a good idea to put the main calleable logic first and helper functions below them